pwiz_tools/Skyline/Model/Results/ChromatogramCache.cs

```diff
@@ -1342,7 +1441,16 @@ public IEnumerable<ChromKeyIndices> GetChromKeys(MsDataFileUri msDataFilePath)
                         int start = lastEntry.StartTransitionIndex;
                         int end = start + lastEntry.NumTransitions;
                         for (int j = start; j < end; j++)
-                            listKeepTransitions.Add(_rawData.ChromTransitions[j]);
+                        {
+                            var chromTransition = _rawData.ChromTransitions[j];
+                            if (formatVersion < CacheFormatVersion.Seventeen && chromTransition.OptimizationStep != 0)
+                            {
+                                // Convert to old format with shifted m/z.
+                                chromTransition.Product += chromTransition.OptimizationStep * ChromatogramInfo.OPTIMIZE_SHIFT_SIZE;

```

Comment:
Could you make a method on ChromTransition that returns a new ChromTransition with the Product and Optimization Step set to the correct value.
It would be nice if ChromTransition could remain immutable so that people don't accidentally modify instances that don't belong to them.

Reply:
I agree, though currently it isn't actually immutable. The flags have setters, e.g. in `ChromCacheBuilder.WriteChromDataSet` there is this:

`chromTran.MissingMassErrors = true;`